### ISSUE TITLE
2 out of 3 members panic when upgrading from 2.0.1 to 2.0.9

### ISSUE THREAD

**Issue by fraenkel** [coreos/etcd#2706]
We were upgrading a 3 node environment.
The first node was successfully upgraded from 2.0.1 -> 2.0.9.

Our other 2 start to panic on start with alternating panics and resolve the issue after about an hour.

```
2015/04/16 21:19:04 etcdserver: start member 16ddfd8ab9399a99 in cluster 47021f879c1614df
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 became follower at term 0
2015/04/16 21:19:04 raft: newRaft 16ddfd8ab9399a99 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 became follower at term 1
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 [term: 1] received a MsgHeartbeat message with higher term from 942b5bcc165d3d6f [term: 151]
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 became follower at term 151
2015/04/16 21:19:04 raft.node: 16ddfd8ab9399a99 elected leader 942b5bcc165d3d6f at term 151
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 [logterm: 0, index: 33133105] rejected msgApp [logterm: 151, index: 33133105] from 942b5bcc165d3d6f
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 [commit: 0, lastindex: 0, lastterm: 0] starts to restore snapshot [index: 33123669, term: 151]
2015/04/16 21:19:04 raftlog: log [committed=1, applied=0, unstable.offset=0, len(unstable.Entries)=0] starts to restore snapshot [index: 33123669, term: 151]
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 restored progress of 942b5bcc165d3d6f [next = 33123670, match = 0, wait = 0]
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 restored progress of dbac133a02199046 [next = 33123670, match = 0, wait = 0]
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 restored progress of f56437f52767f9ea [next = 33123670, match = 0, wait = 0]
2015/04/16 21:19:04 raft: 16ddfd8ab9399a99 [commit: 0] restored snapshot [index: 33123669, term: 151]
2015/04/16 21:19:04 etcdserver: saved incoming snapshot at index 33123669
2015/04/16 21:19:04 rafthttp: starting client stream to 942b5bcc165d3d6f at term 151
2015/04/16 21:19:04 etcdserver: recovered from incoming snapshot at index 33123669
2015/04/16 21:19:05 raft: 16ddfd8ab9399a99 [commit: 33123669] ignored snapshot [index: 33123669, term: 151]
2015/04/16 21:19:05 rafthttp: unexpected removal of unknown peer '15829047930164711494'
2015/04/16 21:19:05 rafthttp: encountered error reading the client log stream: read tcp 10.10.116.10:7001: use of closed network connection
2015/04/16 21:19:05 rafthttp: client streaming to 942b5bcc165d3d6f at term 151 has been stopped
panic: rafthttp: unexpected removal of unknown peer '15829047930164711494'

goroutine 19 [running]:
log.Panicf(0x8d53f0, 0x31, 0xc208d354f8, 0x1, 0x1)
	/var/vcap/data/packages/golang_1.4/80fd10aaca7fafada6e7f228f03c5024faa0b51a.1-9d3a1c8e5be38f64424db093270d5ab90f5ed290/src/log/log.go:314 +0xd0
github.com/coreos/etcd/rafthttp.(*transport).removePeer(0xc20808b1f0, 0xdbac133a02199046)
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/transport.go:152 +0x1c3
github.com/coreos/etcd/rafthttp.(*transport).RemovePeer(0xc20808b1f0, 0xdbac133a02199046)
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/transport.go:136 +0x77
```

```
2015/04/16 21:20:14 etcdserver: start member 16ddfd8ab9399a99 in cluster 47021f879c1614df
2015/04/16 21:20:14 raft: 16ddfd8ab9399a99 became follower at term 0
2015/04/16 21:20:14 raft: newRaft 16ddfd8ab9399a99 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/04/16 21:20:14 raft: 16ddfd8ab9399a99 became follower at term 1
2015/04/16 21:20:14 raft: 16ddfd8ab9399a99 [term: 1] received a MsgHeartbeat message with higher term from 942b5bcc165d3d6f [term: 151]
2015/04/16 21:20:14 raft: 16ddfd8ab9399a99 became follower at term 151
2015/04/16 21:20:14 tocommit(33133112) is out of range [lastIndex(0)]
panic: tocommit(33133112) is out of range [lastIndex(0)]

goroutine 10 [running]:
log.Panicf(0x8c9670, 0x2c, 0xc2080ca510, 0x2, 0x2)
	/var/vcap/data/packages/golang_1.4/80fd10aaca7fafada6e7f228f03c5024faa0b51a.1-9d3a1c8e5be38f64424db093270d5ab90f5ed290/src/log/log.go:314 +0xd0
github.com/coreos/etcd/raft.(*raftLog).commitTo(0xc208038b90, 0x1f99238)
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/log.go:177 +0x179
github.com/coreos/etcd/raft.(*raft).handleHeartbeat(0xc20803a580, 0x8, 0x16ddfd8ab9399a99, 0x942b5bcc165d3d6f, 0x97, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:593 +0x47
github.com/coreos/etcd/raft.stepFollower(0xc20803a580, 0x8, 0x16ddfd8ab9399a99, 0x942b5bcc165d3d6f, 0x97, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:563 +0xfdd
github.com/coreos/etcd/raft.(*raft).Step(0xc20803a580, 0x8, 0x16ddfd8ab9399a99, 0x942b5bcc165d3d6f, 0x97, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:461 +0x23f
github.com/coreos/etcd/raft.(*node).run(0xc208038b40, 0xc20803a580)
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:275 +0x689
created by github.com/coreos/etcd/raft.StartNode
	/var/vcap/data/compile/etcd/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:168 +0x50a
```

It does eventually start to work, and I don't really know why...
```
2015/04/16 21:21:24 etcdserver: start member 1548e6e4f941098d in cluster 47021f879c1614df
2015/04/16 21:21:24 raft: 1548e6e4f941098d became follower at term 0
2015/04/16 21:21:24 raft: newRaft 1548e6e4f941098d [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/04/16 21:21:24 raft: 1548e6e4f941098d became follower at term 1
2015/04/16 21:21:24 raft: 1548e6e4f941098d [term: 1] received a MsgHeartbeat message with higher term from 942b5bcc165d3d6f [term: 151]
2015/04/16 21:21:24 raft: 1548e6e4f941098d became follower at term 151
2015/04/16 21:21:24 raft.node: 1548e6e4f941098d elected leader 942b5bcc165d3d6f at term 151
2015/04/16 21:21:24 raft: 1548e6e4f941098d [logterm: 0, index: 33133815] rejected msgApp [logterm: 151, index: 33133815] from 942b5bcc165d3d6f
2015/04/16 21:21:24 raft: 1548e6e4f941098d [commit: 0, lastindex: 0, lastterm: 0] starts to restore snapshot [index: 33133682, term: 151]
2015/04/16 21:21:24 raftlog: log [committed=1, applied=0, unstable.offset=0, len(unstable.Entries)=0] starts to restore snapshot [index: 33133682, term: 151]
2015/04/16 21:21:24 raft: 1548e6e4f941098d restored progress of 942b5bcc165d3d6f [next = 33133683, match = 0, wait = 0]
2015/04/16 21:21:24 raft: 1548e6e4f941098d restored progress of a3871bb6d0a075ef [next = 33133683, match = 0, wait = 0]
2015/04/16 21:21:24 raft: 1548e6e4f941098d [commit: 0] restored snapshot [index: 33133682, term: 151]
2015/04/16 21:21:24 etcdserver: saved incoming snapshot at index 33133682
2015/04/16 21:21:24 rafthttp: starting client stream to 942b5bcc165d3d6f at term 151
2015/04/16 21:21:24 etcdserver: recovered from incoming snapshot at index 33133682
2015/04/16 21:21:25 raft: 1548e6e4f941098d [commit: 33133682] ignored snapshot [index: 33133682, term: 151]
2015/04/16 21:21:25 etcdserver: added local member 1548e6e4f941098d [http://10.10.115.10:7001] to cluster 47021f879c1614df
2015/04/16 21:21:25 etcdserver: published {Name:etcd_z2-0 ClientURLs:[http://10.10.115.10:4001]} to cluster 47021f879c1614df
2015/04/16 21:30:47 rafthttp: encountered error reading the client log stream: read tcp 10.10.116.10:7001: use of closed network connection
2015/04/16 21:30:47 rafthttp: client streaming to 942b5bcc165d3d6f at term 151 has been stopped
```

**Comment by yichengq** [comment]
`2015/04/16 21:20:14 etcdserver: start member 16ddfd8ab9399a99 in cluster 47021f879c1614df`
Based on this log, i think these two members start without the original data dir.

I tried to reproduce it(start 2.0.1 cluster, shut down, restart them with 2.0.9), and it works fine for me.

**Comment by fraenkel** [comment]
Our upgrade doesn't quite work that way.
We have 3 up, we bring one down, upgrade, bring it back up and then continue to the next member.

**Comment by xiang90** [comment]
@fraenkel 
For some reason, one of your upgraded member does not have the previous wal and snapshot files. We are pretty sure this happened to your cluster based on the log you provided. This is a unsafe behavior and might lead to data inconsistency.

your log
```
2015/04/16 21:21:24 etcdserver: start member 1548e6e4f941098d in cluster 47021f879c1614df
```

restarting a member with data should look like this
```
2015/04/17 19:39:03 etcdserver: restart member 6e3bd23ae5f1eae0 in cluster e9c7614f68f35fb2 at commit index 71
```

The leader of raft was doing its best to prevent re-adding a member who lost its data. The error gets resolved when the leader switched to a new one, who is unaware of the the data loss of the member.

**Comment by xiang90** [comment]
> We have 3 up, we bring one down, upgrade, bring it back up and then continue to the next member.

Rolling upgrade is good. But I would suggest you to check the upgraded member is OK before upgrading another one.

**Comment by fraenkel** [comment]
So what we were doing is removing the current member, then wiping the data_dir. After the code was upgraded, we added the member, and waited until a successful sync.

We did learn that if we do the above without removing the data_dir, we do not panic.  Just seems a bit confusing since we remove the member which is what is causing the panic when we add it back without a data_dir.

**Comment by xiang90** [comment]
@fraenkel 
> So what we were doing is removing the current member, then wiping the data_dir. After the code was upgraded, we added the member, and waited until a successful sync.

You do not need to remove the data-dir. And if you lose the data-dir, it means you lose the member. You have to remove the member and add it back again via run time configuration change API.

Please see https://github.com/coreos/etcd/blob/master/Documentation/admin_guide.md#lifecycle.

**Comment by jpalermo** [comment]
The way the machines are created and destroyed requires us to remove the member then add it back.

If we remove the member, wipe the data_dir, then add the member, when the server starts we get the follow error about `rafthttp: unexpected removal of unknown peer '4954212033102115228'`

If we then try to start again, it seems to come up just fine.

But if we remove the member, then add the member back again, the server produces the same error each time we do this cycle.

```
2015/04/21 18:21:30 etcd: listening for peers on http://0.0.0.0:7001
2015/04/21 18:21:30 etcd: listening for client requests on http://0.0.0.0:4001
2015/04/21 18:21:30 etcdserver: datadir is valid for the 2.0.1 format
2015/04/21 18:21:30 etcdserver: name = etcd_z1-1
2015/04/21 18:21:30 etcdserver: data dir = /var/vcap/store/etcd
2015/04/21 18:21:30 etcdserver: member dir = /var/vcap/store/etcd/member
2015/04/21 18:21:30 etcdserver: heartbeat = 50ms
2015/04/21 18:21:30 etcdserver: election = 1000ms
2015/04/21 18:21:30 etcdserver: snapshot count = 10000
2015/04/21 18:21:30 etcdserver: advertise client URLs = http://10.10.16.35:4001
2015/04/21 18:21:30 etcdserver: start member c92a4990fad5c84a in cluster f61148e36337417a
2015/04/21 18:21:30 raft: c92a4990fad5c84a became follower at term 0
2015/04/21 18:21:30 raft: newRaft c92a4990fad5c84a [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/04/21 18:21:30 raft: c92a4990fad5c84a became follower at term 1
2015/04/21 18:21:30 raft: c92a4990fad5c84a [term: 1] received a MsgHeartbeat message with higher term from c49b621b412fb83e [term: 74]
2015/04/21 18:21:30 raft: c92a4990fad5c84a became follower at term 74
2015/04/21 18:21:30 raft.node: c92a4990fad5c84a elected leader c49b621b412fb83e at term 74
2015/04/21 18:21:30 raft: c92a4990fad5c84a [logterm: 0, index: 478105] rejected msgApp [logterm: 74, index: 478105] from c49b621b412fb83e
2015/04/21 18:21:30 raft: c92a4990fad5c84a [logterm: 0, index: 478105] rejected msgApp [logterm: 74, index: 478105] from c49b621b412fb83e
2015/04/21 18:21:30 raft: c92a4990fad5c84a [commit: 0, lastindex: 0, lastterm: 0] starts to restore snapshot [index: 470068, term: 74]
2015/04/21 18:21:30 raftlog: log [committed=1, applied=0, unstable.offset=0, len(unstable.Entries)=0] starts to restore snapshot [index: 470068, term: 74]
2015/04/21 18:21:30 raft: c92a4990fad5c84a restored progress of 44c0e5987704719c [next = 470069, match = 0, wait = 0]
2015/04/21 18:21:30 raft: c92a4990fad5c84a restored progress of 602db4bc0b4b8334 [next = 470069, match = 0, wait = 0]
2015/04/21 18:21:30 raft: c92a4990fad5c84a restored progress of c49b621b412fb83e [next = 470069, match = 0, wait = 0]
2015/04/21 18:21:30 raft: c92a4990fad5c84a [commit: 0] restored snapshot [index: 470068, term: 74]
2015/04/21 18:21:30 etcdserver: saved incoming snapshot at index 470068
2015/04/21 18:21:30 raft: c92a4990fad5c84a [commit: 470068] ignored snapshot [index: 470068, term: 74]
2015/04/21 18:21:30 rafthttp: starting client stream to c49b621b412fb83e at term 74
2015/04/21 18:21:30 etcdserver: recovered from incoming snapshot at index 470068
2015/04/21 18:21:31 raft: c92a4990fad5c84a [commit: 470068] ignored snapshot [index: 470068, term: 74]
2015/04/21 18:21:31 rafthttp: unexpected removal of unknown peer '4954212033102115228'
2015/04/21 18:21:31 rafthttp: encountered error reading the client log stream: read tcp 10.10.80.19:7001: use of closed network connection
2015/04/21 18:21:31 rafthttp: client streaming to c49b621b412fb83e at term 74 has been stopped
panic: rafthttp: unexpected removal of unknown peer '4954212033102115228'

goroutine 19 [running]:
log.Panicf(0x8d6d90, 0x31, 0xc2083734f8, 0x1, 0x1)
        /var/vcap/data/packages/golang1.4/f57ddbc8d55d7a0f08775bf76bb6a27dc98c7ea7.1-b7b27b40f381b6fb59e45e37ea96847381781996/src/log/log.go:314 +0xd0
github.com/coreos/etcd/rafthttp.(*transport).removePeer(0xc2080831f0, 0x44c0e5987704719c)
        /var/vcap/data/compile/etcd/etcd/gopath/src/github.com/coreos/etcd/rafthttp/transport.go:152 +0x1c3
github.com/coreos/etcd/rafthttp.(*transport).RemovePeer(0xc2080831f0, 0x44c0e5987704719c)
        /var/vcap/data/compile/etcd/etcd/gopath/src/github.com/coreos/etcd/rafthttp/transport.go:136 +0x77
github.com/coreos/etcd/etcdserver.(*EtcdServer).applyConfChange(0xc20806e200, 0x9c4cdcffc2a1dcca, 0x1, 0x44c0e5987704719c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /var/vcap/data/compile/etcd/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:791 +0x81e
github.com/coreos/etcd/etcdserver.(*EtcdServer).apply(0xc20806e200, 0xc208526048, 0x1f84, 0x1f8d, 0xc20803ddd0, 0x0, 0x0)
        /var/vcap/data/compile/etcd/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:699 +0x399
github.com/coreos/etcd/etcdserver.(*EtcdServer).run(0xc20806e200)
        /var/vcap/data/compile/etcd/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:430 +0xf4c
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).start
        /var/vcap/data/compile/etcd/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:298 +0x294
```

**Comment by xiang90** [comment]
> remove the member, wipe the data_dir, then add the member

How did you remove the member and add the member back?

**Comment by jpalermo** [comment]
etcdctl -C '${CLUSTER_ENDPOINTS}' member remove ${my_id}

etcdctl -C '${CLUSTER_ENDPOINTS}' member add ${node_name} ${advertise_peer_url}

**Comment by xiang90** [comment]
@jpalermo I think this is related to https://github.com/coreos/etcd/pull/2701. 

It will be resolved in 2.1 soon. And we will back-port the patch back to 2.0 release if possible.
